### PR TITLE
[Bugfix][ROCm] Force splitK=0 in AiterFp8BlockScaledMMKernel for determinism

### PR DESCRIPTION
**Pull request by maeehart**
## Summary

`aiter.gemm_a8w8_blockscale_ck` is non-deterministic when its CSV-driven
dispatcher selects `splitK >= 2`: the underlying CK split-K reduction
accumulates partial K products through atomic floating-point adds,
which is not associative, so back-to-back calls with bit-identical
inputs can disagree by a few LSBs in bf16. This PR forces
`splitK=0` on the CK path inside
`AiterFp8BlockScaledMMKernel.apply_block_scaled_mm` so the kernel is
deterministic, bypassing the buggy CSV dispatcher until the underlying
CK reduction is fixed upstream.

This is intentionally a small, scoped fix.

## Reproducer (12 lines, single GPU, MI355X / gfx950)

```python
import aiter, torch
from vllm.model_executor.layers.quantization.utils.fp8_utils \
    import per_token_group_quant_fp8
M, K, N = 4, 4096, 7168                         # o_proj at decode batch 4
A_bf = torch.randn(M, K, device="cuda", dtype=torch.bfloat16)
Aq, As = per_token_group_quant_fp8(A_bf, 128)
B  = torch.randn(N, K, device="cuda", dtype=torch.bfloat16).clamp_(-3, 3)
Bs = torch.full((N // 128, K // 128), 1.0, device="cuda")
Bq = B.to(torch.float8_e4m3fn)
Y0 = torch.zeros(M, N, device="cuda", dtype=torch.bfloat16)
Y1 = torch.zeros(M, N, device="cuda", dtype=torch.bfloat16)
aiter.gemm_a8w8_blockscale_ck(Aq, Bq, As, Bs, Y0, splitK=3)  # CSV pick
aiter.gemm_a8w8_blockscale_ck(Aq, Bq, As, Bs, Y1, splitK=3)
print((Y0 - Y1).abs().max().item())                           # ~2-4  (NOT 0!)
```

| splitK | max\|Y0-Y1\| (M=4, N=7168, K=4096, bf16) |
|-------:|------------------------------------------:|
| 0      | 0.0                                       |
| 2      | ~2.5                                      |
| 3      | ~2.0–4.0                                  |

Verified on the patched build:

```
splitK=3 nondeterminism: max|d|= 2.0
splitK=0 max|d|= 0.0
```

## When did it start?

`aiter` PR
[#3024](https://github.com/ROCm/aiter/pull/3024) ("[Silo] Add configs
missing from bulk merge #3004", merged 2026-05-05) added 147 `gfx950`
rows to
`aiter/configs/model_configs/a8w8_blockscale_tuned_gemm_ds_v3.csv`,
replacing previously deterministic `splitK=0` entries with
`splitK={2,3}` for some DSV3.2 dense decode shapes. The re-roll was a
pure speed optimisation — the tuner is not aware of the determinism
cost of split-K. `git blame` confirms PR #3024 (`1638f9e`) is the
trigger.

## Scope of the fix

Only the CK path of `AiterFp8BlockScaledMMKernel` is changed. The
Triton path (`triton_gemm_a8w8_blockscale`) already uses a deterministic
two-stage reduction (no atomics) and is left untouched. The CK-Tile
path is also bypassed — it consumes the same `splitK` from the same
CSV and uses the same atomic-FP-add reduction, so it is affected by
the same bug. Routing all non-Triton traffic through CK at `splitK=0`
keeps the workaround uniformly deterministic and minimal.

## Restriction to 128x128 weight blocks

`aiter.gemm_a8w8_blockscale_ck` is hardcoded for 128x128 weight
blocks. To make this assumption explicit and prevent silent misuse,
`can_implement` now refuses to bind to any layer whose
`weight_quant_key.scale.group_shape != GroupShape(128, 128)`, and
`apply_block_scaled_mm` asserts the same invariant defensively at the
call site. The kernel is currently only ever bound to (128, 128)
weights in tree, so this is a no-op for existing configurations.

## Verification

End-to-end on `vllm-project/vllm@215e2f799` + the team's standard ROCm
flags (`VLLM_ROCM_USE_AITER=1`,
`VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1`,
`VLLM_ROCM_QUICK_REDUCE_QUANTIZATION=INT4`,
`--tensor-parallel-size 4 --block-size 64 --async-scheduling
--gpu-memory-utilization 0.85 --enforce-eager --max-num-seqs 4`),
DeepSeek-V3.2, MI355X TP4:

| benchmark                         | un-patched | patched | delta   |
|-----------------------------------|-----------:|--------:|--------:|
| gsm8k 5-shot, c=4, limit=200, flex | 0.925     | 0.930   | +0.005  |
| gsm8k 5-shot, c=4, limit=200, strict | 0.925   | 0.930   | +0.005  |
| stderr (both rows)                | 0.0187     | 0.0181  | —       |
| Output throughput (tok/s, c=64, ISL=1024, OSL=256) | 231.87 | 236.52 | +2.0% |
| TPOT mean (ms, same)              | 211.98     | 206.91  | -2.4%   |
| TPOT median (ms, same)            | 229.07     | 225.45  | -1.6%   |
| TTFT median (ms, same)            | 3804.06    | 3997.99 | +5.1%   |

Accuracy is unchanged within stderr; performance is within run-to-run
noise. The forced `splitK=0` is concentrated on small-M decode shapes
where split-K's theoretical benefit was already small.

## Follow-up upstream

A draft AITER patch (clamp `splitK<=1` in `aiter/ops/gemm_op_a8w8.py`
plus `op_test_gemm_a8w8_blockscale_determinism.py`) is ready to send to
ROCm/aiter. Once that lands, this vLLM workaround can be removed.

### CODE REVIEW

File: vllm/model_executor/kernels/linear/scaled_mm/aiter.py [hunk 1]
```diff
@@ -328,6 +357,18 @@ def apply_block_scaled_mm(
...
+        if _AITER_FORCE_SPLITK_ZERO and not self.use_triton:
+            # Bypass the CSV-driven splitK selection in
+            # `aiter.gemm_a8w8_blockscale` and call the CK kernel directly
+            # with splitK=0. See module-level comment for rationale.
+            import aiter
+
+            M = A.shape[0]
+            N = B.shape[0]
+            Y = torch.empty(M, N, dtype=out_dtype, device=A.device)
+            return aiter.gemm_a8w8_blockscale_ck(A, B, As, Bs, Y, splitK=0)
```
[the remark this answers is not in the document]
Comment (maeehart) [reply vllm/model_executor/kernels/linear/scaled_mm/aiter.py]: Thanks for catching this. The framing is slightly off — neither the dispatcher path nor the direct call ever passes `block_size` down to AITER (the upstream `aiter.gemm_a8w8_blockscale` signature doesn't accept one, and the vLLM wrapper `rocm_aiter_ops.gemm_a8w8_blockscale` accepts a `block_size: list[int]` arg but silently drops it before calling AITER) — so switching to the direct CK call doesn't actually lose any block-size routing.

However, the underlying concern is valid: `can_implement` previously only constrained the **activation** group_shape, not the weight group_shape, which means a non-128x128 weight block would silently bind to this kernel and call into a kernel that's hardcoded for 128x128. I've addressed that more strictly than your suggestion in the latest commit:

1. `can_implement` now refuses any `weight_quant_key.scale.group_shape != GroupShape(128, 128)`, so the dispatcher won't even pick this kernel for unsupported shapes (instead of falling through to the still-broken-for-non-128x128 dispatcher path, which would silently produce wrong results).
2. Added a defensive `assert` at the call site documenting the same invariant.
3. Updated the module-level comment to also mention the CK-Tile path is bypassed for the same reason (it shares the atomic-FP-add reduction).
4. Removed the env-var opt-out — there's no legitimate reason to opt back into the non-deterministic kernel from production code.

PTAL.

**Comment by tjtanaa** [comment]
@maeehart we are now at aiter v0.1.13, has that been fixed in aiter v0.1.13? On top of this, is there any AITER tracking issue number?

**Comment by maeehart** [comment]
I have to do more testing and see if we should fix this on the AITER side or if it has already been fixed. Let me make this into a draft.

**Comment by maeehart** [comment]
Superseded by [ROCm/aiter#4581](https://github.com/ROCm/aiter/pull/4581). The AITER fix retains CSV-selected CK and CKTile kernel dispatch, validates the requested `splitK` exponent, and clamps only exponents greater than 1 at both compiled entry points. Direct regressions cover CK and CKTile requests for exponents 2 and 3, and rebuilt modules passed all four cases on MI355X.

Closing this workaround in favor of fixing the root cause in AITER.
